### DISCUSSION TITLE
BUILD FAILED (macOS 13.5.1 using ruby-build 20230914.1)

### DISCUSSION THREAD

**Discussion by thomasdiben** [rbenv/ruby-build#2253] [Build failures] [answered]
Hi everyone!
I'm trying to deploy my react native app to TestFlight using fastlane and it seems I need to be using ruby 2.7.8 for this.
Thus my presence here: after following the install/setup instructions, I get stuck after trying `rbenv install 2.7.8`

Here's what I got from the console:
```
ruby-build: using readline from homebrew
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew

BUILD FAILED (macOS 13.5.1 using ruby-build 20230914.1)

Inspect or clean up the working tree at /var/folders/8p/5lrfwn5s29s05mmnsy64789r0000gn/T/ruby-build.20230918122229.13695.QKHKGO
Results logged to /var/folders/8p/5lrfwn5s29s05mmnsy64789r0000gn/T/ruby-build.20230918122229.13695.log

Last 10 log lines:
In file included from vm.c:350:
./vm_method.c:291:32: warning: cast to smaller integer type 'enum method_optimized_type' from 'void *' [-Wvoid-pointer-to-enum-cast]
            def->body.optimize_type = (enum method_optimized_type)opts;
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 warnings and 1 error generated.
make: *** [vm.o] Error 1
make: *** Waiting for unfinished jobs....
1 warning generated.
1 warning generated.
1 warning generated.
```
I did update ruby-build already. Anyone can help on this?
Thanks a lot!

**Comment by mislav** [comment]
Hi, not sure what went wrong, but the output doesn't show the actual failure message. If you open the full build log (observe the `Results logged to` line) and search for keywords like "error", you might find the actual error message. Note that Ruby 2.7 is "end-of-life" and has stopped receiving updates, which means that it could have problems installing on recent architectures and newer versions of macOS. Good luck!

**Reply by thomasdiben** [reply]
Thanks @mislav for your answer! And you're absolutely right, I juste checked the log and we have more information there. Nothing I can make sense of unfortunately:
```
Configuration summary for ruby version 2.7.8

   * Installation prefix: /Users/thomas/.rbenv/versions/2.7.8
   * exec prefix:         ${prefix}
   * arch:                x86_64-darwin22
   * site arch:           ${arch}
   * RUBY_BASE_NAME:      ruby
   * enable shared:       yes
   * ruby lib prefix:     ${libdir}/${RUBY_BASE_NAME}
   * site libraries path: ${rubylibprefix}/${sitearch}
   * vendor path:         ${rubylibprefix}/vendor_ruby
   * target OS:           darwin22
   * compiler:            clang
   * with pthread:        yes
   * with coroutine:      amd64
   * enable shared libs:  yes
   * dynamic library ext: bundle
   * CFLAGS:              ${optflags} ${debugflags} ${warnflags}
   * LDFLAGS:             -L. -L/Users/thomas/.rbenv/versions/2.7.8/lib  \
                          -fstack-protector-strong -L/usr/local/lib
   * DLDFLAGS:            -L/Users/thomas/.rbenv/versions/2.7.8/lib  \
                          -Wl,-multiply_defined,suppress
   * optflags:            -O3
   * debugflags:          -ggdb3
   * warnflags:           -Wall -Wextra -Wdeprecated-declarations \
                          -Wdivision-by-zero \
                          -Wimplicit-function-declaration -Wimplicit-int \
                          -Wmisleading-indentation -Wpointer-arith \
                          -Wshorten-64-to-32 -Wwrite-strings \
                          -Wmissing-noreturn -Wno-cast-function-type \
                          -Wno-constant-logical-operand -Wno-long-long \
                          -Wno-missing-field-initializers \
                          -Wno-overlength-strings -Wno-parentheses-equality \
                          -Wno-self-assign -Wno-tautological-compare \
                          -Wno-unused-parameter -Wno-unused-value \
                          -Wunused-variable -Wextra-tokens
   * strip command:       strip -A -n
   * install doc:         rdoc
   * JIT support:         yes
   * man page type:       doc
   * BASERUBY -v:         ruby 2.6.10p210 (2022-04-12 revision 67958) \
                          [universal.x86_64-darwin22]

```
And here's the error:
```
In file included from vm.c:344:
./vm_insnhelper.c:5045:43: error: use of undeclared identifier 'RUBY_FUNCTION_NAME_STRING'
        fprintf(stderr, "%s %s(%d):%p\n", RUBY_FUNCTION_NAME_STRING, bf->name, bf->argc, bf->func_ptr);
                                          ^
compiling vm_backtrace.c
compiling vm_dump.c
In file included from vm.c:350:
./vm_method.c:124:5: warning: function 'rb_f_notimplement' declared 'noreturn' should not return [-Winvalid-noreturn]
    UNREACHABLE_RETURN(Qnil);
    ^
./include/ruby/ruby.h:59:35: note: expanded from macro 'UNREACHABLE_RETURN'
#  define UNREACHABLE_RETURN(val) return (val)
                                  ^
In file included from vm.c:350:
./vm_method.c:291:32: warning: cast to smaller integer type 'enum method_optimized_type' from 'void *' [-Wvoid-pointer-to-enum-cast]
            def->body.optimize_type = (enum method_optimized_type)opts;
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 warnings and 1 error generated.
make: *** [vm.o] Error 1
make: *** Waiting for unfinished jobs....
1 warning generated.
1 warning generated.
1 warning generated.
```
Thanks again 🙏

**Reply by mislav** [reply, marked as answer]
Have you tried installing or upgrading Xcode command-line tools? Either `xcode-select --install` (for first-time install) or use Software Update to upgrade an already existing installation.

Another thing you can try is `RUBY_CONFIGURE_OPTS="--disable-shared" rbenv install ...`. If your invocation line already has RUBY_CONFIGURE_OPTS (for example, to link to a specific openssl version), then just add `--disable-shared` to the existing list of options. Good luck!

Possibly ref. https://github.com/ruby/ruby/commit/31bc55dc90aa18222d5ed968e62fea4d46554d8c

**Reply by thomasdiben** [reply]
Thanks @mislav 
I've tried your suggestions and I've been able to install 2.7.8. Thanks a lot! 🙏 
For me, updating xcode (14->15) did the job.

**Reply by ah3243** [reply]
god i hate ruby. I just spend an hour installing different packages each with different compatibilities.

**Comment by monfresh** [comment]
@thomasdiben Are you on an Apple Silicon Mac by any chance? If so, the problem could be that you're using Rosetta or running commands with `arch -x86_64`. If so, I recommend uninstalling Ruby and starting over in native mode. When you run `arch`, it should say `arm64`.

Also, try installing Ruby 3.2.2 or 3.1.4. If it doesn't work, then the problem is most likely with something missing, broken, outdated, or misconfigured in your dev setup. `brew doctor` is a good place to start to check for issues. Another common source of issues is your shell startup files (.zshrc, .zprofile, .zlogin, etc). Check them for any Ruby compilation or configuration flags and remove/disable them.

If you can install Ruby 3.2.2 or 3.1.4, then I recommend upgrading your React Native project from 2.7.8 to either 3.1.4 or 3.2.2. For the most part, it should only require replacing "2.7.8" in the `.ruby-version` and `Gemfile` files. For a more detailed step-by-step tutorial, check out this article I wrote:

https://www.rubyonmac.dev/how-to-upgrade-the-ruby-version-in-your-project

**Reply by thomasdiben** [reply]
You rock, @monfresh thanks for the thorough answer. It was indeed an outdated dev setup (xcode.. Which I was reluctant to update).
Edit: I have an Intel chip
Edit2: interestingly enough, I had already read your article 👍
